pkg/services/grpcserver/auth.go

```diff
@@ -0,0 +1,128 @@
+package grpcserver
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/grafana/grafana/pkg/cmd/grafana-cli/logger"
+	apikeygenprefix "github.com/grafana/grafana/pkg/components/apikeygenprefixed"
+	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/models"
+	"github.com/grafana/grafana/pkg/services/apikey"
+	"github.com/grafana/grafana/pkg/services/sqlstore"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/status"
+)
+
+// Authenticator can authenticate GRPC requests.
+type Authenticator struct {
+	logger   log.Logger
+	APIKey   apikey.Service
+	SQLStore sqlstore.Store
+}
+
+func NewAuthenticator(apiKey apikey.Service, sqlStore sqlstore.Store) *Authenticator {
+	return &Authenticator{
+		logger:   log.New("grpc-server-authenticator"),
+		APIKey:   apiKey,
+		SQLStore: sqlStore,
+	}
+}
+
+// Authenticate checks that a token exists and is valid. It stores the user
+// metadata in the returned context and removes the token from the context.
+func (a *Authenticator) Authenticate(ctx context.Context) (context.Context, error) {
```

Comment:
do we need to export it? it looks like a grpc server's internal middleware

Reply:
switched to lower case in https://github.com/grafana/grafana/pull/47849/commits/7ed4f39219b839d3046008352e11a8019a200ce8